sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dofn/InitializeDoFn.java

```diff
@@ -53,6 +52,25 @@ public void processElement(OutputReceiver<com.google.cloud.Timestamp> receiver)
     LOG.info(daoFactory.getStreamTableDebugString());
     LOG.info(daoFactory.getMetadataTableDebugString());
     LOG.info("ChangeStreamName: " + daoFactory.getChangeStreamName());
+    if (!daoFactory
+        .getMetadataTableAdminDao()
+        .isAppProfileSingleClusterAndTransactional(this.metadataTableAppProfileId)) {
+      LOG.error(
+          "App profile id '"
+              + metadataTableAppProfileId
+              + "' provided to access metadata table needs to use single-cluster routing policy"
+              + " and allow single-row transactions.");
+      // Terminate this pipeline now.
+      return;
```

Comment:
I suggest performing this check in the `expand` method - this would allow you to fail the pipeline submission and have an error thrown.

When you perform the check at this point in the pipeline, the pipeline will terminate, but it will look 'normal' - users will need to notice the ERROR log.

We can leave this check here, but I think you should make sure to perform the same check in the `expand` method so the pipeline submission fails with an error being thrown (rather than just logged)

Reply:
That's a good suggestion. I think fail fast is a good idea. The biggest worry that I have is the pipeline submitter might not have the necessary permissions to interact with Cloud Bigtable. Whereas the service account attached to the dataflow workers will. 

I think I will need some extra time to evaluate this approach. So for now, I think we should go ahead for now without checking in `expand`.